### PR TITLE
feat: Add cache control for jwks endpoint

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Add cache control for jwks endpoint

## Related issues

-   https://github.com/supertokens/supertokens-node/pull/668

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.
-   [ ] If access token structure has changed
    -   Modified test in `tests/sessions/test_access_token_version.py` to account for any new claims that are optional or omitted by the core

### CODE REVIEW

File: supertokens_python/querier.py [hunk 1]
```diff
@@ -259,10 +259,17 @@ async def __send_request_helper(
...
+                res["_text"] = response.text
+
             try:
-                return response.json()
+                res.update(response.json())
             except JSONDecodeError:
-                return response.text
+                pass
+
+            return res
```
Comment (KShivendu) [inline supertokens_python/querier.py:272]: I tried with creating a QuerierResponse class as well but that will require too many changes. Plus, we aren't actually using response.text anywhere. So made it part of the dict response.

File: supertokens_python/recipe/jwt/interfaces.py [hunk 2]
```diff
@@ -40,8 +40,9 @@ class CreateJwtResultUnsupportedAlgorithm:
 
 
 class GetJWKSResult:
-    def __init__(self, keys: List[JsonWebKey]):
+    def __init__(self, keys: List[JsonWebKey], validity_in_secs: Optional[int]):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/jwt/interfaces.py:43]: validity_in_secs does not need to be optional

File: supertokens_python/recipe/jwt/api/implementation.py [hunk 3]
```diff
@@ -25,4 +25,10 @@ async def jwks_get(
         self, api_options: APIOptions, user_context: Dict[str, Any]
     ) -> JWKSGetResponse:
         response = await api_options.recipe_implementation.get_jwks(user_context)
+
+        if response.validity_in_secs is not None:
+            api_options.response.set_header(
+                "Cache-Control", f"max-age={response.validity_in_secs}, must-revalidate"
+            )
+
         return JWKSGetResponse(response.keys)
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/jwt/api/implementation.py:34]: write an extra test wherein you query the jwks endpoint and check that resulting json. Make sure the resulting json only contains the keys props with the right values. No extra props should be there.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/jwt/interfaces.py (hunk 2 above)
Comment (KShivendu) [reply supertokens_python/recipe/jwt/interfaces.py:43]: Not doing this based on our discussion on slack.

File: supertokens_python/recipe/jwt/api/implementation.py (hunk 3 above)
Comment (KShivendu) [reply supertokens_python/recipe/jwt/api/implementation.py:34]: Done

File: supertokens_python/querier.py [hunk 4]
```diff
@@ -285,10 +285,17 @@ async def __send_request_helper(
...
+
+            if response.headers.get("content-type", "").startswith("text"):
+                res["_text"] = response.text
+
             try:
-                return response.json()
+                res.update(response.json())
             except JSONDecodeError:
-                return response.text
+                pass
```
Comment (rishabhpoddar) [inline supertokens_python/querier.py:296]: this is changing the semantics of how things were done. Please fix

File: supertokens_python/recipe/jwt/interfaces.py (hunk 2 above)
Comment (rishabhpoddar) [inline supertokens_python/recipe/jwt/interfaces.py:43]:
```suggestion
    def __init__(self, keys: List[JsonWebKey], validity_in_secs: Optional[int] = None):
```

File: supertokens_python/recipe/jwt/interfaces.py [hunk 5]
```diff
@@ -40,8 +40,9 @@ class CreateJwtResultUnsupportedAlgorithm:
 
 
 class GetJWKSResult:
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/jwt/interfaces.py:42]: double check where all this is used, and you need to pass in a validity_in_secs in all the use of it. It should only be created in the recipe interface.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/querier.py (hunk 4 above)
Comment (KShivendu) [reply supertokens_python/querier.py:296]: Done

File: supertokens_python/recipe/jwt/interfaces.py (hunk 2 above)
Comment (KShivendu) [reply supertokens_python/recipe/jwt/interfaces.py:43]: Done

File: supertokens_python/recipe/jwt/interfaces.py (hunk 5 above)
Comment (KShivendu) [reply supertokens_python/recipe/jwt/interfaces.py:42]: Done

**Comment by KShivendu** [comment]
We need to rebase against 0.16 before merging.

**Pull request merged**